File: backend/tools/websearch.py
from mixedbread import Mixedbread


class WebSearch:
    """Web search tool using mixedbread's web store."""

    def __init__(self, api_key: str):
        self.mixedbread = Mixedbread(api_key=api_key)
        self.store_name = "mixedbread/web"

    def search(self, query: str, top_k: int = 10) -> list[dict]:
        """
        Search the web using mixedbread's web store.

        Args:
            query: Search query
            top_k: Number of results to return
        """
        resp = self.mixedbread.stores.search(
            store_identifiers=[self.store_name],
            query=query,
            top_k=top_k,
        )
        results = []
        for item in resp.data:
            content = item.text if hasattr(item, "text") else ""
            metadata = item.metadata if hasattr(item, "metadata") else {}

            title = metadata.get("title", "")
            url = metadata.get("url", "") or (item.filename if hasattr(item, "filename") else "")

            results.append({
                "title": title or url,
                "url": url,
                "content": content,
                "score": item.score if hasattr(item, "score") else 0,
            })
        return results

    @staticmethod
    def get_tool() -> list[dict]:
        """Returns tool descriptions for OpenAI Responses API function calling."""
        return [
            {
                "type": "function",
                "name": "web_search",
                "description": "Search the web for current information. Use this when you need up-to-date information, facts, or to research topics beyond your knowledge.",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "query": {
                            "type": "string",
                            "description": "The search query",
                        },
                        "top_k": {
                            "type": "integer",
                            "description": "Number of results to return (default: 10)",
                        },
                    },
                    "required": ["query"],
                },
            },
        ]

File: ui/lib/services/web-search.ts
import Mixedbread from "@mixedbread/sdk";

export interface WebSearchResult {
  title: string;
  url: string;
  content: string;
  score: number;
}

export class WebSearch {
  private client: Mixedbread;
  private storeName = "mixedbread/web";

  constructor(apiKey: string) {
    this.client = new Mixedbread({ apiKey });
  }

  async search(query: string, topK = 10): Promise<WebSearchResult[]> {
    const resp = await this.client.stores.search({
      store_identifiers: [this.storeName],
      query,
      top_k: topK,
    });

    return (resp.data ?? []).map((item) => {
      // Handle different chunk types
      let content = "";
      if ("text" in item && typeof item.text === "string") {
        content = item.text;
      }

      const metadata = (item.metadata as Record<string, unknown>) ?? {};
      const title = (metadata.title as string) ?? "";
      const url = (metadata.url as string) ?? item.filename ?? "";

      return {
        title,
        url,
        content,
        score: item.score ?? 0,
      };
    });
  }
}
